### PR TITLE
Cache results of tree ecobenefit summary searches

### PR DESCRIPTION
**Pull request by RickMohr**
* Check cache before computing tree benefit summary
* Update cache after computing tree benefit summary
* Cache key is `eco/trees/{url_name}/{eco_rev}/{filter_string}/{display_string}`
* Timeout for cache entries is one month
* Use Redis cache

Update `eco_rev`:
* Anywhere `geo_rev` is updated
* After a species import (a species' i-Tree region may have changed)

`update_map_feature()`
* Simplify `geo_rev` checking -- is field a subclass of `GeometryField`?
* Add a test which creates and updates a Bioswale
* Add eco_rev checking
* Update/add tests so `eco_rev` cases are covered

Connects #2260
Connects #2256

### CODE REVIEW

**Comment by jwalgran** [comment]
Looking at this now. Travis has an eco-related test failure:

```
======================================================================
ERROR: test_benefit_search (treemap.tests.test_urls.TreemapUrlTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/OpenTreeMap/otm-core/opentreemap/treemap/tests/test_urls.py", line 242, in test_benefit_search
    'treemap/partials/eco_benefits.html')
  File "/home/travis/build/OpenTreeMap/otm-core/opentreemap/treemap/tests/test_urls.py", line 41, in assert_template
    response = self.assert_status_code(url, 200)
  File "/home/travis/build/OpenTreeMap/otm-core/opentreemap/treemap/tests/test_urls.py", line 29, in assert_status_code
    response = send(url, data)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/test/client.py", line 500, in get
    **extra)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/test/client.py", line 303, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/test/client.py", line 379, in generic
    return self.request(**r)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/test/client.py", line 466, in request
    six.reraise(*exc_info)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/travis/build/OpenTreeMap/otm-core/opentreemap/treemap/decorators.py", line 45, in wrapper
    return view_fn(request, instance, *args, **kwargs)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/views/decorators/http.py", line 158, in inner
    response = func(request, *args, **kwargs)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django_tinsel/decorators.py", line 79, in wrapper
    params = callable_or_dict(request, *args, **wrapper_kwargs)
  File "/home/travis/build/OpenTreeMap/otm-core/opentreemap/treemap/views/tree.py", line 68, in search_tree_benefits
    benefits, basis = get_benefits_for_filter(filter)
  File "/home/travis/build/OpenTreeMap/otm-core/opentreemap/treemap/ecobenefits.py", line 416, in get_benefits_for_filter
    ft_benefit_groups, ft_basis = _benefits_for_class(C, filter)
ValueError: too many values to unpack

**Comment by jwalgran** [comment]
I am seeing the benefit search test failure on my dev VM too. I am also seeing a failure in the new bioswale test

```
ERROR: test_create_bioswale (treemap.tests.test_map_feature.GeoAndEcoRevIncr)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/app/core/opentreemap/treemap/tests/test_map_feature.py", line 55, in test_create_bioswale
    self.instance.add_map_feature_types(['Bioswale'])
  File "/usr/local/lib/python2.7/dist-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/opt/app/core/opentreemap/treemap/instance.py", line 505, in add_map_feature_types
    raise ValidationError('Map feature types already added: %s' % dups)
ValidationError: [u"Map feature types already added: set([u'Bioswale'])"]

**Comment by jwalgran** [comment]
I am seeing the effects of the cache.

I am using this helper shell command to clear out any existing eco cache for an instance:

```
redis-cli -n 1 keys ":1:eco/trees/largeinstance/*" | grep ":1:eco/trees/largeinstance/.*" | xargs redis-cli -n 1 DEL
```

Unfiltered and filtered requests are faster with cached values.

```
> time curl -I "http://localhost:6060/largeinstance/benefit/search"
real	0m8.130s
$ time curl -I "http://localhost:6060/largeinstance/benefit/search"
real	0m3.686s

$ time curl -I "http://localhost:6060/largeinstance/benefit/search?q=%7B%22tree.diameter%22%3A%7B%22MIN%22%3A%220%22%2C%22MAX%22%3A%221002%22%7D%7D"
real	0m6.062s
$ time curl -I "http://localhost:6060/largeinstance/benefit/search?q=%7B%22tree.diameter%22%3A%7B%22MIN%22%3A%220%22%2C%22MAX%22%3A%221002%22%7D%7D"
real	0m1.751s
```

Immediately trashing the cache proves that it is providing the speed up.

```
$ time curl -I "http://localhost:6060/largeinstance/benefit/search"
real	0m7.702s
$ redis-cli -n 1 keys ":1:eco/trees/largeinstance/*" | grep ":1:eco/trees/largeinstance/.*" | xargs redis-cli -n 1 DEL
$ time curl -I "http://localhost:6060/largeinstance/benefit/search"
real	0m8.776s

File: requirements.txt [hunk 1]
```diff
@@ -26,3 +26,4 @@ python-dateutil==2.2
 pytz==2014.7
 django-tinsel==0.1.1
 django-url-tools==0.0.8
+django-redis-cache==1.6.0
```
Comment (jwalgran) [inline requirements.txt]: We have more experience with ``django-redis`` on other projects. When quickly writing up issue #2256 I referenced the wrong cache interface library. The Django cache framework abstracts this library choice, so switching in ``django-redis`` should be relatively transparent.

**Comment by jwalgran** [comment]
The migration is not backwards compatible with old code, which we are trying to get in the habit of doing in order to ease deployments.

screen shot 2015-09-21 at 12 28 45 pm

**Comment by RickMohr** [comment]
#2274 is why it's still too slow. We should fix that.

**Comment by RickMohr** [comment]
@jwalgran, all fixed. I couldn't replicate the Bioswale test failure.

**Comment by jwalgran** [comment]
I verified

* The caching of benefits works as before with the new library
* The new setting flag does disable caching
* I could successfully create an instance with the old code after applying this migration

On previous projects we have use the the [hiredis](https://pypi.python.org/pypi/hiredis) parser and wired it in under ``OPTIONS`` like this:

```python
CACHES = {
    'default': {
        ...
        'OPTIONS': {
            ...
            'PARSER_CLASS': 'redis.connection.HiredisParser',
        }
    }
}
```

The [django-redis docs](http://niwinz.github.io/django-redis/latest/#_pluggable_parsers) say that this gives a "performance boost" 

The benchmarks on the [hiredis-python README](https://github.com/redis/hiredis-py#benchmarks)  show the biggest gains in LRANGE performance.

The hiredis package is pretty lightweight and pip compiles/installs it quickly

```
$ time sudo pip install hiredis
...
real	0m1.345s
```

I am inclined to install it while were are setting up our caching, but I don't have any performance benchmarks to show an improvement for our immediate usage pattern.

:+1:

**Comment by RickMohr** [comment]
@jwalgran, I added HiRedis, verified it worked, and checked in the family. When we roll out those LRANGE operations we'll be blazin'.

**Pull request merged**